greedy_algorithms/gale_shapley.cpp

```diff
@@ -0,0 +1,112 @@
+ * 
+ * Reference: 
+ * [Wikipedia](https://en.wikipedia.org/wiki/Gale%E2%80%93Shapley_algorithm)
+ * [Wikipedia](https://en.wikipedia.org/wiki/Stable_matching_problem)
+ *
+ * @author [B Karthik](https://github.com/BKarthik7)
+ */
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+#include <cassert>
+
+/**
+ * @namespace
+ * @brief Greedy Algorithms
+ */
+namespace greedy_algorithms {
+/**
+ * @namespace
+ * @brief Functions for the Gale-Shapley Algorithm
+ */
+namespace stable_matching {
+std::vector<int> gale_shapley(const std::vector<std::vector<int>>& secondary_preferences, const std::vector<std::vector<int>>& primary_preferences) {
+    int num_elements = secondary_preferences.size();
+    std::vector<int> matches(num_elements, -1);
+    std::vector<bool> is_free_primary(num_elements, true);
+    std::vector<int> proposal_index(num_elements, 0); // Tracks the next secondary to propose for each primary
+
+    while (true) {
+        int free_primary_index = -1;
+
+        // Find the next free primary
+        for (int i = 0; i < num_elements; i++) {
+            if (is_free_primary[i]) {
+                free_primary_index = i;
+                break;
+            }
+        }
+
+        // If no free primary is found, break the loop
+        if (free_primary_index == -1) break;
+
+        // Get the next secondary to propose
+        int secondary_to_propose = primary_preferences[free_primary_index][proposal_index[free_primary_index]];
+        proposal_index[free_primary_index]++;
+
+        // Get the current match of the secondary
+        int current_match = matches[secondary_to_propose];
+
+        // If the secondary is free, match them
+        if (current_match == -1) {
+            matches[secondary_to_propose] = free_primary_index;
+            is_free_primary[free_primary_index] = false;
+        } else {
+            // Determine if the current match should be replaced
+            auto new_proposer_rank = std::find(secondary_preferences[secondary_to_propose].begin(),
+                                                secondary_preferences[secondary_to_propose].end(),
+                                                free_primary_index);
+            auto current_match_rank = std::find(secondary_preferences[secondary_to_propose].begin(),
+                                                 secondary_preferences[secondary_to_propose].end(),
+                                                 current_match);
+
+            // If the new proposer is preferred over the current match
+            if (new_proposer_rank < current_match_rank) {
+                matches[secondary_to_propose] = free_primary_index;
+                is_free_primary[free_primary_index] = false;
+                is_free_primary[current_match] = true; // Current match is now free
+            }
+        }
+    }
+
+    return matches;
+}
+}  // namespace stable_matching
+}  // namespace greedy_algorithms
+
+
+static void tests() {
```

Comment:
Tests function should be documented.

Reply:
Can you explain what documentation mean?